### PR TITLE
Delete rows or columns

### PR DESCRIPTION
**Pull request by Lunalo**
updated the dialog by Editing  SetDefaults and SetRcode for controls 

@africanmathsinitiative/developers  this is ready for review

### CODE REVIEW

**Comment by shadrackkibet** [comment]
I am going to test this now.

**Comment by lilyclements** [comment]
I'm making minor edits to this dialog now, make sure you've sent all your commits :)

File: instat/dlgDeleteRowsOrColums.vb [hunk 1]
```diff
@@ -67,17 +66,23 @@ Public Class dlgDeleteRowsOrColums
...
-        SetRCode(Me, ucrBase.clsRsyntax.clsBaseFunction, bReset)
+        ucrSelectorForDeleteColumns.AddAdditionalCodeParameterPair(clsDeleteColumns, ucrSelectorForDeleteColumns.GetParameter, iAdditionalPairNo:=1)
+        ucrPnlColumnsOrRows.SetRCode(clsDeleteRows)
+        ucrPnlColumnsOrRows.SetRCode(clsDeleteColumns)
+        ucrNudRowsToDelete.SetRCode(clsDeleteRows)
+        ucrReceiverForColumnsToDelete.SetRCode(clsDeleteColumns)
+        ucrSelectorForDeleteColumns.SetRCode(clsDeleteRows)
     End Sub
-
     Private Sub TestOKEnabled()
```
Comment (lilyclements) [inline instat/dlgDeleteRowsOrColums.vb]: Line 80 - you removed a blank line. Can you put this blank line back in? It's good to have a space between subs.

**Comment by shadrackkibet** [comment]
i think we have a problem in deleting all columns.The dialog breaks.Is there a way you can solve this @Lunalo ?

File: instat/dlgDeleteRowsOrColums.vb [hunk 2]
```diff
@@ -67,17 +66,23 @@ Public Class dlgDeleteRowsOrColums
...
+        clsDeleteColumns.SetRCommand(frmMain.clsRLink.strInstatDataObject & "$remove_columns_in_data")
+        ucrBase.clsRsyntax.SetBaseRFunction(clsDeleteColumns)
     End Sub
 
     Private Sub SetRCodeForControls(bReset As Boolean)
-        SetRCode(Me, ucrBase.clsRsyntax.clsBaseFunction, bReset)
+        ucrSelectorForDeleteColumns.AddAdditionalCodeParameterPair(clsDeleteColumns, ucrSelectorForDeleteColumns.GetParameter, iAdditionalPairNo:=1)
+        ucrPnlColumnsOrRows.SetRCode(clsDeleteRows)
+        ucrPnlColumnsOrRows.SetRCode(clsDeleteColumns)
+        ucrNudRowsToDelete.SetRCode(clsDeleteRows)
```
Comment (lilyclements) [inline instat/dlgDeleteRowsOrColums.vb]: This ucrNud does not reset if you press reset. You haven't set it a default value - you've only said what the `RDefault `is.

**Comment by lilyclements** [comment]
@shadrackkibet does it crash? We have a problem in general on R-Instat that it will crash if you remove all the columns.

**Comment by rdstern** [comment]
Good to have found this problem.  Remember the Danny rule that we want dialogues to be used simply, without restrictions, unless it causes a real problem.

If you want to remove all columns or rows, I assume you are effectively deleting the sheet.  We do have a dialogue specially for this.  So we could allow it from here, or we could trap that situation and give a nice message that directs users to the "proper way of doing this.

What we mustn't have is R-Instat crashing like this, so great the problem has been found.

**Comment by shadrackkibet** [comment]
yes it crashes. I had raised this issue earlier  at  #2981 which danny closed.

**Comment by shadrackkibet** [comment]
could there be an option for deleting more than one row at the same time on this dialog?

image

**Comment by Lunalo** [comment]
@dannyparsons  @rdstern 

I would like to implement the functionality suggested by @shadrackkibet 

R allows us to have dataframe with 0 rows and 0 columns. 
For our case If I open empty dataframe  then I should have columns with NAs in the cells.

@shadrackkibet  why would you delete all the columns or rows rows? I think it is easier for you to delete the sheet but not  all the columns nor rows. 

But I can provide a check for empty column list and dataframe to prevent this error.

File: instat/dlgDeleteRowsOrColums.vb [hunk 3]
```diff
@@ -58,24 +57,30 @@ Public Class dlgDeleteRowsOrColums
...
+        clsDeleteRows.SetRCommand(frmMain.clsRLink.strInstatDataObject & "$remove_rows_in_data")
+        clsDeleteColumns.SetRCommand(frmMain.clsRLink.strInstatDataObject & "$remove_columns_in_data")
+        ucrBase.clsRsyntax.SetBaseRFunction(clsDeleteColumns)
     End Sub
 
     Private Sub SetRCodeForControls(bReset As Boolean)
-        SetRCode(Me, ucrBase.clsRsyntax.clsBaseFunction, bReset)
+        ucrSelectorForDeleteColumns.AddAdditionalCodeParameterPair(clsDeleteColumns, ucrSelectorForDeleteColumns.GetParameter, iAdditionalPairNo:=1)
+        ucrPnlColumnsOrRows.SetRCode(clsDeleteRows)
+        ucrPnlColumnsOrRows.SetRCode(clsDeleteColumns)
```
Comment (lilyclements) [inline instat/dlgDeleteRowsOrColums.vb]: This panel has had the r-code set twice for two different classes. This isn't the correct way to do this perhaps have a look at a frequency dialog to look into how this should be set for a panel with multiple base functions/classes

**Comment by lilyclements** [comment]
@Lunalo I don't fully understand.
You said:

> R allows us to have dataframe with 0 rows and 0 columns.
> For our case If I open empty dataframe then I should have columns with NAs in the cells.

I don't understand, if it's an empty dataframe with just NAs, how do we know how many columns/rows to give it?
Or, if it has no columns and no rows, we can't put NAs in, it's just a blank sheet.


To the second point, I don't see why this dialog shouldn't allow the user to delete all their columns or rows if they wanted to. I know it's not as simple or easy as deleting the sheet, but I don't think it should be disallowed just because there's an alternative way to do it.

**Comment by lilyclements** [comment]
Finally, we do have the capacity to delete multiple rows, it just hasn't been implemented in this dialog.
If you try in R, then it does work:

e.g.
`InstatDataObject$remove_rows_in_data(data_name="Survey", row_names=c(1:10))
`
will delete rows 1 to 10.

I'm not sure why this hasn't yet been implemented? Perhaps @dannyparsons can explain that one.

I suppose it is difficult to say which rows to remove, but we can always only allow the user to delete rows if they're after each other, i.e.,

Delete Rows
From: NUD    To: NUD

**Comment by shadrackkibet** [comment]
I don't have a perfect reason to support why the user should be allowed to delete all the rows/columns.But my concern is why should you limit the user not to do this on this dialog because of another alternative way?

**Comment by dannyparsons** [comment]
This is a good discussion. The issue on deleting all columns was closed because it had already been written up somewhere else. It is still something we must fix. This in in the R code so should be automatically solved for this dialog.

In general R does support 0 rows and/or 0 columns so we should try to support this. However if it's really difficult for us (e.g. grid can't easily support 0 rows/columns) and then I don't think it's a problem if we don't allow this in R-Instat.

I like the suggestions on improvements to the dialog. Let's also not get too distracted from the main task of updating to the new system. However, I think the change of deleting **From row x To y** should be simple and would be a great improvement. If this is done with an ROperator then it should work well in with the new system. And also `row_names = 1:10` would do the same as `row_names = 10:1` so there shouldn't be additional restrictions needed.

**Comment by maxwellfundi** [comment]
@dannyparsons  we have now added the operators for this dialog. @Lunalo the nuds seem to go down to zero, which causes a bug. Once that is fixed this will be ready

**Comment by lilyclements** [comment]
@Lunalo How have you managed to only edit one file? Are all your commits sent?
This should be at least two files: Designer code and VB-dialog-code

**Comment by Lunalo** [comment]
@lilyclements 

Thanks for that and sorry for the delay , I was having a session with the students on R-Instat this afternoon

**Comment by lilyclements** [comment]
@Lunalo it's okay, I was just confused how it happened. Hope the session went well!

**Comment by lilyclements** [comment]
@Lunalo I made some minor edits which are waiting to be merged

This dialog should be called dlgDeleteRowsOrColumns not dlgDeleteRowsOrColums (missing a "n").

Is it possible to correct this dialog name?
Just a warning: I have asked Steven to change a dialog name on #3099 and I don't know if changing two names may create a conflict so check with him before you do if you do! We can always do it another time if not.

**Comment by Lunalo** [comment]
@lilyclements 

I have merged your edits. I think the dialog is in a good shape

**Comment by lilyclements** [comment]
@Lunalo you still need to address the point on line 94 where you've set the r code twice for the same panel. Look at other dialogs - such as frequencies - on how this is tackled

**Comment by Lunalo** [comment]
@lilyclements 

I think I gave attention to something else and forgot to solve that

**Comment by lilyclements** [comment]
@dannyparsons this is ready for you to review

File: instat/dlgDeleteRowsOrColums.vb [hunk 4]
```diff
@@ -57,25 +58,43 @@ Public Class dlgDeleteRowsOrColums
...
         ucrSelectorForDeleteColumns.Reset()
-        clsDefaultFunction.SetRCommand(frmMain.clsRLink.strInstatDataObject & "$remove_columns_in_data")
-        ucrBase.clsRsyntax.SetBaseRFunction(clsDefaultFunction.Clone())
+
+        clsDeleteRows.SetRCommand(frmMain.clsRLink.strInstatDataObject & "$remove_rows_in_data")
+        clsDeleteColumns.SetRCommand(frmMain.clsRLink.strInstatDataObject & "$remove_columns_in_data")
+
+        clsOperator.SetOperation(":")
+        clsOperator.AddParameter("From", 1, iPosition:=0)
+        clsOperator.AddParameter("To", 1, iPosition:=2)
```
Comment (Lunalo) [inline instat/dlgDeleteRowsOrColums.vb]: I think 1 is for the operation itself

File: instat/dlgDeleteRowsOrColums.vb [hunk 5]
```diff
@@ -99,7 +118,8 @@ Public Class dlgDeleteRowsOrColums
     Private Sub ReopenDialog()
         If ucrSelectorForDeleteColumns.ucrAvailableDataFrames.cboAvailableDataFrames.Text <> "" Then
             Try
-                ucrNudRowsToDelete.Maximum = frmMain.clsRLink.GetDataFrameLength(ucrSelectorForDeleteColumns.ucrAvailableDataFrames.cboAvailableDataFrames.Text)
+                ucrNudFrom.Maximum = frmMain.clsRLink.GetDataFrameLength(ucrSelectorForDeleteColumns.ucrAvailableDataFrames.cboAvailableDataFrames.Text)
+                ucrNudTo.Maximum = frmMain.clsRLink.GetDataFrameLength(ucrSelectorForDeleteColumns.ucrAvailableDataFrames.cboAvailableDataFrames.Text)
```
Comment (Lunalo) [inline instat/dlgDeleteRowsOrColums.vb]: I didnt want to get rid of this before because I was unsure. I will have a look at it

Comment (Lunalo) [inline instat/dlgDeleteRowsOrColums.vb]: this is now fixed

File: instat/dlgDeleteRowsOrColums.vb (hunk 4 above)
Comment (Lunalo) [inline instat/dlgDeleteRowsOrColums.vb]: I have fixed it

File: instat/dlgDeleteRowsOrColums.vb [hunk 6]
```diff
@@ -14,12 +14,12 @@
 ' You should have received a copy of the GNU General Public License k
 ' along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-Imports instat
 Imports instat.Translations
 Public Class dlgDeleteRowsOrColums
     Public bFirstLoad As Boolean = True
     Private bReset As Boolean = True
-
+    Private clsOperator As New ROperator
```
Comment (Lunalo) [inline instat/dlgDeleteRowsOrColums.vb]: This is fixed

**Comment by Lunalo** [comment]
@dannyparsons 

This is now ready

**Comment by Lunalo** [comment]
@africanmathsinitiative/developers 

can someone test this please

**Comment by lilyclements** [comment]
This isn't running in the `data_name` for me?

`InstatDataObject$remove_columns_in_data(cols="Size", data_name=)`

**Comment by lilyclements** [comment]
This `data_name =` issue was an issue in a few dialogs which Danny has now fixed. Retested the dialog and @dannyparsons this is ready to review.

**Pull request merged**